packages/angular/src/wallet/wallet.store.ts

```diff
@@ -0,0 +1,295 @@
+                        tap(() => {
+                            localStorage.setItem(this._localStorageKey, walletName);
+                            const wallet = wallets.find(({ name }) => name === walletName);
+                            const adapter = wallet ? wallet.adapter() : null;
+                            this.patchState({
+                                selectedWallet: walletName as WalletName,
+                                adapter,
+                                wallet,
+                                ready: adapter?.ready || false,
+                            });
+                        })
+                    )
+            )
+        );
+    });
+
+    readonly onConnect = this.effect(() => {
+        return this.adapter$.pipe(
+            isNotNull,
+            switchMap((adapter) =>
+                of(adapter).pipe(
+                    fromAdapterEvent('connect'),
+                    tap(() =>
+                        this.patchState({
+                            connected: true,
+                            publicKey: adapter.publicKey,
+                            autoApprove: adapter.autoApprove,
+                        })
+                    )
+                )
+            )
+        );
+    });
+
+    readonly onDisconnect = this.effect(() => {
+        return this.adapter$.pipe(
+            isNotNull,
+            fromAdapterEvent('disconnect'),
+            tap(() =>
+                this.patchState({
+                    connected: false,
+                    connecting: false,
+                    disconnecting: false,
+                    autoApprove: false,
+                    publicKey: null,
+                })
+            )
+        );
+    });
+
+    readonly onReady = this.effect(() => {
+        return this.adapter$.pipe(
+            isNotNull,
+            fromAdapterEvent('ready'),
+            tap(() => this.patchState({ ready: true }))
+        );
+    });
+
+    readonly onError = this.effect(() => {
+        return this.adapter$.pipe(
+            isNotNull,
+            fromAdapterEvent('error'),
+            tap((error) => this.logError(error))
+        );
+    });
+
+    sendTransaction(
+        transaction: Transaction,
+        connection: Connection,
+        options?: SendTransactionOptions
+    ): Observable<string> {
+        return this.state$.pipe(
+            first(),
+            concatMap(({ adapter, connected }) => {
+                if (!adapter) {
+                    return throwError(new WalletNotSelectedError());
+                }
+
+                if (!connected) {
+                    return throwError(new WalletNotConnectedError());
+                }
+
+                return from(defer(() => adapter.sendTransaction(transaction, connection, options))).pipe(
+                    map((txId) => txId as string)
+                );
+            })
+        );
+    }
+
+    signTransaction(transaction: Transaction): Observable<Transaction> | undefined {
```

Comment:
How would this be called by a dapp, and how does the dapp detect whether the adapter allows it? Just trying to understand the type safety this provides.

Reply:
Unlike `Promises` that are eager, `Observables` are lazy, they do nothing until you `subscribe`. Even if you call `signTransaction`, nothing happens until you subscribe. Given that this can return undefined, you can't simply subscribe without checking first.

A common dapp usage would have to look like:

```typescript
const signTransaction$ = this.walletStore.signTransaction(tx);

if (!signTransaction$) {
  notifyError()
}

signTransaction$.subscribe()
```

Without the check, the signTransaction$ can be undefined and wouldn't have the subscribe method. Given the `adapter && 'signTransaction' in adapter` check, the adapter type inferred is `SignerWalletAdapter`, and for  `signMessage` it's `MessageSignerWalletAdapter`.

Reply:
Interesting, okay. Makes sense then.